### PR TITLE
feat(application-system): Add notifications support to pruning

### PR DESCRIPTION
**Pull request by norda-gunni**
# ...

Attach a link to issue if relevant

## What

Add the ability for applications to send notifications to users when their applications are pruned

## Why

For instance a user who hasn't completed an application might be confused when an application they hadn't completed disappeared.

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
	- Enhanced user notification capabilities integrated into the application lifecycle.
	- New module for eager user notifications added, allowing for improved notification handling.
	- Expanded application data returned, including new fields for better application management.

- **Bug Fixes**
	- Improved error handling for notification sending processes.

- **Documentation**
	- Added section on configuring custom pruning notifications in the README.

- **Tests**
	- Introduced unit tests for notification handling in the application lifecycle service.

### CODE REVIEW

File: libs/application/templates/README.md [hunk 1]
```diff
@@ -1,5 +1,33 @@
...
+For example an application might want to send a notification when an application that was in the draft stage was pruned. Here is an example of how that might be accimplished in the application template:
+
+```typescript
+stateMachineConfig: {
+    initial: States.DRAFT,
+    states: {
+      [States.DRAFT]: {
+        meta: {
+          name: application.general.name.defaultMessage,
+          progress: 0.4,
```
Comment (obmagnusson) [inline libs/application/templates/README.md:18]:
```suggestion
 
```

File: libs/application/templates/README.md [hunk 2]
```diff
@@ -1,5 +1,33 @@
 # Templates
 
+## Pruning notifications
+
+If desired it is possible to have a custom message sent to an application's applicant when it is pruned.
+This is done by adding a `pruneMessage` object of type `PruningNotification` or a function returning a `PruningNotification` to the application's lifecycle configuration.
+The function will be passed an argument containing a `PruningApplication` object which contains the application's data. The `PruningNotification` object has `externalBody`, `internalBody` and `notificationTemplateId`. The former two can be used to fill in the values for the `externalBody` and `internalBody` template variables in the notification that will be sent to the user. The `notificationTemplateId` is the id of the notification template that will be used to send the notification. It is up to the user whether to use the body variables or not.
+
+For example an application might want to send a notification when an application that was in the draft stage was pruned. Here is an example of how that might be accimplished in the application template:
```
Comment (obmagnusson) [inline libs/application/templates/README.md:9]:
```suggestion
For example an application might want to send a notification when an application that was in the draft stage was pruned. Here is an example of how that might be accomplished in the application template:
```

File: libs/application/templates/README.md [hunk 3]
```diff
@@ -1,5 +1,33 @@
 # Templates
 
+## Pruning notifications
+
+If desired it is possible to have a custom message sent to an application's applicant when it is pruned.
+This is done by adding a `pruneMessage` object of type `PruningNotification` or a function returning a `PruningNotification` to the application's lifecycle configuration.
+The function will be passed an argument containing a `PruningApplication` object which contains the application's data. The `PruningNotification` object has `externalBody`, `internalBody` and `notificationTemplateId`. The former two can be used to fill in the values for the `externalBody` and `internalBody` template variables in the notification that will be sent to the user. The `notificationTemplateId` is the id of the notification template that will be used to send the notification. It is up to the user whether to use the body variables or not.
```
Comment (obmagnusson) [inline libs/application/templates/README.md:7]:
```suggestion
The function will be passed as an argument containing a `PruningApplication` object which contains the application's data. The `PruningNotification` object has `externalBody`, `internalBody` and `notificationTemplateId`. The former two can be used to fill in the values for the `externalBody` and `internalBody` template variables in the notification that will be sent to the user. The `notificationTemplateId` is the id of the notification template that will be used to send the notification. It is up to the user whether to use the body variables or not.
```

File: libs/application/templates/README.md [hunk 4]
```diff
@@ -1,5 +1,33 @@
...
+
+```typescript
+stateMachineConfig: {
+    initial: States.DRAFT,
+    states: {
+      [States.DRAFT]: {
+        meta: {
+          name: application.general.name.defaultMessage,
+          progress: 0.4,
+          lifecycle: defaultLifecycleWithPruneMessage((application: any) => ({
```
Comment (obmagnusson) [inline libs/application/templates/README.md:19]: This should be typed ?

File: libs/application/templates/README.md (hunk 2 above)
Comment (norda-gunni) [reply libs/application/templates/README.md:9]: Good catch, thanks.

File: libs/application/templates/README.md (hunk 3 above)
Comment (norda-gunni) [reply libs/application/templates/README.md:7]: The intention here was to say that when the function gets executed it gets passed an argument that is a PruningApplication object. 
I have updated the relevant text to try to make that clearer, thanks for pointing this out.

File: libs/application/templates/README.md (hunk 4 above)
Comment (norda-gunni) [reply libs/application/templates/README.md:19]: Indeed, thanks for catching that.

**Review by obmagnusson — APPROVED** [review]

**Review by thordurhhh — APPROVED** [review]

**Pull request merged**